collector/perf_schema_file_instances.go

```diff
@@ -0,0 +1,96 @@
+// Scrape `performance_schema.file_summary_by_instance`.
+
+package collector
+
+import (
+	"database/sql"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"flag"
+	"strings"
+)
+
+const perfFileInstancesQuery = `
+	SELECT
+	    FILE_NAME,
+	    COUNT_READ, COUNT_WRITE,
+	    SUM_NUMBER_OF_BYTES_READ, SUM_NUMBER_OF_BYTES_WRITE
+	  FROM performance_schema.file_summary_by_instance
+	     where FILE_NAME REGEXP ?
+	`
+
+// Metric descriptors.
+var (
+	performanceSchemaFileInstancesFilter = flag.String(
+		"collect.perf_schema.file_instances.filter", ".*",
+		"RegEx file_name filter for performance_schema.file_summary_by_instance",
+	)
+
+	performanceSchemaFileInstancesRemovePrefix = flag.Bool(
+		"collect.perf_schema.file_instances.remove_prefix", true,
+		"Remove path prefix in performance_schema.file_summary_by_instance",
+	)
+
+	performanceSchemaFileInstancesBytesDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, performanceSchema, "file_instances_bytes"),
+		"The number of bytes processed by file read/write operations.",
+		[]string{"file_name", "mode"}, nil,
+	)
+	performanceSchemaFileInstancesCountDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, performanceSchema, "file_instances_total"),
+		"The total number of file read/write operations.",
+		[]string{"file_name", "mode"}, nil,
+	)
+)
+
+// ScrapePerfFileEvents collects from `performance_schema.file_summary_by_event_name`.
+func ScrapePerfFileInstances(db *sql.DB, ch chan<- prometheus.Metric) error {
+	// Timers here are returned in picoseconds.
+	perfSchemaFileInstancesRows, err := db.Query(perfFileInstancesQuery, *performanceSchemaFileInstancesFilter)
+	if err != nil {
+		return err
+	}
+	defer perfSchemaFileInstancesRows.Close()
+
+	var (
+		fileName                      string
+		countRead, countWrite         uint64
+		sumBytesRead, sumBytesWritten uint64
+	)
+
+	for perfSchemaFileInstancesRows.Next() {
+		if err := perfSchemaFileInstancesRows.Scan(
+			&fileName,
+			&countRead, &countWrite,
+			&sumBytesRead, &sumBytesWritten,
+		); err != nil {
+			return err
+		}
+
+		if *performanceSchemaFileInstancesRemovePrefix && *performanceSchemaFileInstancesFilter != ".*" {
+			pos := strings.LastIndexAny(fileName, "/\\")
```

Comment:
Yes, this would be more elegant. But it doesn't work if the exporter runs on Linux and MySQL runs on Windows. Then you would have to replace all separator chars of the MySQL Host OS with the exporter Host OS ones first. I don't know a more elegant solution than this:

    p := "c:\\windows\\myfile"
    var other string
	this := strconv.QuoteRune(filepath.Separator)
	if filepath.Separator == '/' {
		other = "\\"
	} else {
		other = "/"
	}
	p = strings.Replace(p, other, this, -1)
	fmt.Printf("Base[%s]\n", filepath.Base(p))

What do you think?